Documentation/building/windows-instructions.md

```diff
@@ -124,6 +124,13 @@ For more details, or to test an individual project, see the [developer guide top
 4. Select the corresponding launch profile (green arrow, i.e. `.NET Core xUnit Console`)
 5. F5 (Debug)
 
+### Using Test Explorer in Visual Studio
+
+1. Open solution from the build script: `.\build.cmd -vs Microsoft.CSharp`. Alternatively you can also pass in the relative or full path to the solution file.
+2. Navigate to the Test Explorer tab and run/debug tests.
+
+VS Test Explorer support is limited to the .NET Core. To switch between Configurations (Debug / Release), Visual Studio needs to be reopened with the command above together with the additional `--configuration/-c` option.
```

Comment:
Why can't we just make it so opening the sln/csproj "just works"?  No one is going to know to do this, and will only figure it out after they fail, ask the question, and we point them to the solution.  This also breaks from the normal flow of cd'ing into a particular directly, doing work there, and opening the solution at some point.

Reply:
Because Visual Studio always looks for the `dotnet` in the PATH. There is currently no way to change that on a repo basis besides setting environment variables.

Reply:
I had an internal discussion with @nguerrera yesterday about this, and his response was that there was pushback on every feature that could make this "just work". corefx requires using a runtime at a specific location, and the Test Explorer only knows how to invoke `dotnet.exe` that is on the `%PATH%`.

Maybe someone on the Test Explorer team can chime in. (I don't know who to @ for that.)

Reply:
Additionally, in the past we talked with @vitek-karas about teaching the host to forward to another host based on a configuration set somewhere, ie global.json but at that point that wasn't feasible because of costs involved with the check. Everyone would pay without play. cc @vitek-karas

Reply:
My initial thinking is that this should be able to be solved with dev tooling, and shouldn't require a runtime change. `launchsettings.json` or similar (like MSBuild properties) should be able to tell the VS Test Explorer where to go looking for `dotnet.exe`.

Reply:
It's nice and I appreciate that you were able to hack up a script-based solution to at least make it feasible for some people.  For me personally, the combination of using a dogfood VS that this apparently won't work with anyway plus this requiring me to fundamentally change my workflow (by the time I realize I want this I will generally already be inside of the project in VS and not at a good point to exit, remember how to relaunch "correctly", etc.).  But I hope it works well for others.

Reply:
> Maybe someone on the Test Explorer team can chime in. (I don't know who to @ for that.)

I created an issue for that quite a while ago: https://github.com/microsoft/vstest/issues/1749 (cc @singhsarab and @mayankbansal018).

Reply:
> launchsettings.json or similar (like MSBuild properties) should be able to tell the VS Test Explorer where to go looking for dotnet.exe.

launchsettings.json is a pure F5 concept and doesn't affect VSTest at all. The pendant would be the runsettings file.

> But I hope it works well for others.

In example: https://github.com/dotnet/corefx/issues/34314#issuecomment-506908769. Supporting VS Test Explorer is one of the most requested features and I'm happy that I can finally enable a solution even if it's a partial one. Getting to this point was a huge amount of work and required a lot of thinking (looking back at the hackathons that @ericstj and I did). There is definitely room for improvement, ie:

- Not having to relaunch VS when switching configurations
- Using the same dotnet host for building and testing
- Not having to use environment variables to point to the repo local dotnet installation

1 & 2 can be fixed with https://github.com/dotnet/core-setup/issues/4809. 3 requires changes in the host that nobody was yet willing to make because of the runtime costs involved with it.

Reply:
> Getting to this point was a huge amount of work and required a lot of thinking (looking back at the hackathons that @ericstj and I did)

Yup!  Like I said, I appreciate all that work, and it's great that there's at least a partial solution in place.

Reply:
> My initial thinking is that this should be able to be solved with dev tooling, and shouldn't require a runtime change.

This would only solve this specific issue with VSTest. The host problem would remain and would continue binding us in unfortunate ways. Let me explain that with one more example: At the moment we are building with from root with the repo local dotnet installation under `.dotnet`. When building individual projects, devs usually use the globally installed dotnet installation. Having different versions of the toolchain is sub-optimal and caused regressions in the past like this one: https://github.com/dotnet/corefx/issues/34455. Teaching the host to forward to our local host is IMHO the only correct solution here that would solve all the mentioned issues.

Reply:
The reasons I think this should be solved by the dev tooling, and not the runtime, are:

1. There hasn't been any scenario where we think this would apply to deployed, production applications. We've only identified cases where developers want this behavior.
2. (more importantly) The security concerns of enabling something like this. Just being able to drop a file somewhere in the directory structure, and now I can change how an app runs is pretty concerning. (The host reads so many files when it launches an application, the performance concern of looking at one more file is not significant to me.)

Reply:
> When building individual projects, devs usually use the globally installed dotnet installation as it's easily invokable (dotnet build / dotnet msbuild). Having different versions of the toolchain is sub-optimal and already caused regressions in the past

There's a flip side to this that gets ignored far too much IMHO. Pinning down the entire toolchain for local development harms dogfooding. What I've seen play out many times with ASP.NET and others that do this is that we get bug reports way too late because the toolchain is not updated frequently enough. If we make it so that everything magically uses a pinned toolchain always, that could get even worse. 

It's interesting to me to think about why we aren't paranoid about pinning down other parts of the toolchain. For example, does anyone worry that we don't all use the same C++ compiler on our machines? I think the answer is that this pinning down is most useful when you have to be on the bleeding edge, as in targeting .NET Core X while also building .NET Core X here. Things are moving quickly and you are likely to be broken often as a result. This is also the exact case where you want the earliest possible signal that things are broken. There's tension between these two goals.

There are other solutions than using global tools and encouraging everyone to dogfood fresh bits globally, which, by the way, has been the division's approach in a world where VS provided the toolset. For example, we can have automated flighting of newer toolchains to all repos. That's  better in many ways, but might not catch everything. Often an official build will work while some VS feature is broken, etc. If we push out dogfooding VS bits that then redirect to something pinned by all of our repos, we won't quite be dogfooding. Today, this is less apparent because VS only uses a sliver of things that we can redirect with hacks like these env vars. If we push for everything to be redirectable, we have to think harder about how this impacts dogfooding.


> Teaching the host to forward to our local host is IMHO the only correct solution here that would solve all the mentioned issues.

There are risks down this path too. As things in VS or other IDEs start to use .NET Core, we have to be sure that they won't be impacted by this. In the time before time, the internal build used a private .NET Framework to host the entire build, and you couldn't launch VS nor a bunch of your favorite tools from the same env. It was a nightmare. In this regard, it may be better to control specific dev actions like launching debug or test processes.

Reply:
> When building individual projects, devs usually use the globally installed dotnet installation as it's easily invokable (dotnet build / dotnet msbuild). 

FWIW: Aside from when building for netstandard2.1 (when it fails and I'm forced to think about it) I use msbuild.  (Mainly because that's what we did in the 1.0-preview days, and it still works, so motor memory wins.)

```
D:\git\bartonjs\corefx\src\System.Security.Cryptography.Pkcs>where msbuild
C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\MSBuild.exe
```

Reply:
> There's a flip side to this that gets ignored far too much IMHO. Pinning down the entire toolchain for local development harms dogfooding. What I've seen play out many times with ASP.NET and others that do this is that we get bug reports way too late because the toolchain is not updated frequently enough. If we make it so that everything magically uses a pinned toolchain always, that could get even worse.

I sympathize with this: I feel the same pain when our up stack partners don't quickly uptake our bits as happens all too often.  The solution to this isn't let things float, but to flow bits more quickly.  We're actually getting there.  Imagine we had back-channel ingestion that forced the new SDK to be picked up on a weekly cadence and didn't require our devs to install dogfood builds machine-wide?  You'd actually get **more** dogfooding, and we'd get deterministic behavior between commandline and VS.

>  does anyone worry that we don't all use the same C++ compiler on our machines?

We would if differences in the C++ toolchain resulted in lost productivity of our developers.  I know WPF folks would very much like this.

> you couldn't launch VS nor a bunch of your favorite tools from the same env
 
That was due to the private GAC, architecture limitations of COMPLUS_ROOT, and version coupling in the mscoreei activation.  Hopefully we can solve those. I agree that we don't want to build using a framework we just built, but I do see value in building with the same framework in VS and the commandline.

I believe @ViktorHofer is adding a stop gap here until we can get "the right" features in either the runtime or tooling to better enable the scenario of an integrated experience using the real VS product to build and test the CoreFx bits.  It's great to see that lots of folks are passionate about this.  Let's treat this as a step in that direction and work on a cohesive plan to address the problem completely in the near-term post 3.0.

Reply:
> C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\MSBuild.exe

smh at 15.0 ;)

Reply:
> Imagine we had back-channel ingestion that forced the new SDK to be picked up on a weekly cadence

I would love to see this. Folks would still have to understand that things will pass through the automated weekly insertion and still impact their productivity when something the official build doesn't use breaks in VS, etc. Using fresh bits, however you get them, will have pain points.  This is life on the edge. As long as we don't abuse our ability to redirect to avoid taking this pain often enough, I'm happy.

> That was due to the private GAC, architecture limitations of COMPLUS_ROOT, and version coupling in the mscoreei activation. Hopefully we can solve those.

Hopefully, indeed! So far DOTNET_ROOT, PATH, and DOTNET_MULTILEVEL lookup are not really all that much better IMHO :( And everything people suggest that doesn't involve such env vars keeps getting rejected. 

> I do see value in building with the same framework in VS and the commandline.

As more parts of VS move to .NET Core, it is not at all clear to me which parts would get value and which parts would just get broken. Having a magic file next to the sln that redirects all child processes of VS doesn't seem like it will work in the long run for this reason.

> I believe @ViktorHofer is adding a stop gap here until we can get "the right" features in either the runtime or tooling

Agree. I wasn't pushing back on this change.

Reply:
Going back a bunch of comments in this thread:

>  the combination of using a dogfood VS that this apparently won't work with anyway

I missed something. Why wouldn't this work when using a dogfood VS?

Reply:
> Yes, it currently doesn't work with the our internal dogfood build.

Don't know... @ViktorHofer wrote "Yes, it currently doesn't work with the our internal dogfood build.".

Reply:
There's a discussion happening now how the VS will run the tests because currently the reliance on `PATH` makes it really hard to correctly run x64 or x86 tests (as the muxer has to be of the right bitness, but `PATH` has no way to specify both really). I raised the concern that ideally there should be a way for the project to specify which muxer should be used to run the tests. It doesn't fully fix the problem described here, but it would definitely help with the test-only problem.

Reply:
> I missed something. Why wouldn't this work when using a dogfood VS?

The latest VS dogfood chooses the wrong TargetFramework when running the tests. @vitek-karas  and I just had a meeting with the VSTest team to let them know.

Additionally we talked about finding the right host for test execution. VSTest currently uses %PATH% to determine the host. Having different host architectures doesn't work well with this. We asked for the following features:
1) Being able to specify a runsettings file in the project file that is used by Test Explorer / dotnet test.
2) to specify the dotnet host that is used for test execution, inside the runsettings file
3) to specify environment variables in the runsettings file

These asks are necessary for both VS Test Explorer and dotnet test to work.

Reply:
Ok. I see. A bug in current dogfood build then. I thought I was hearing that you could never use dogfood with this. 

Is there a tracking issue for the bug? The dogfood bits will be the stable bits very quickly.

Reply:
I expect @mayankbansal018 to let me know what's the best way to follow-up on the issue.

Reply:
Still we want to draft loose requirements for tools/runtime changes to solve the other mentioned issues like using the same host when building inside VS or from command line.